scheduler/remote_server.py

```diff
@@ -0,0 +1,487 @@
+                            else:
+                                # if the prio is lower, then we only the schedule the piece that
+                                # doesn't overlap.
+                                item_to_add['time'] = queued_finish
+                                new_duration = finish_time - queued_finish
+                                item_to_add['duration'] = new_duration.total_seconds()//60
+                                queued_lines.append(item_to_add)
+                    else:
+                        time_diff = item_to_add['time'] - queued_finish
+                        new_line = copy.deepcopy(inf_dur_line)
+                        new_line['duration'] = time_diff.total_seconds()//60
+                        new_line['time'] = queued_finish
+                        queued_lines.append(new_line)
+                        queued_lines.append(item_to_add)
+
+        if idx == len(scd_lines) - 1:
+            last_tt_line, queued_finish = calculate_new_last_line_params()
+
+            inf_dur_line['time'] = queued_finish
+            queued_lines.append(inf_dur_line)
+
+    queued_dict = collections.OrderedDict()
+    for line in queued_lines:
+        if not line['order'] in queued_dict:
+            queued_dict[line['order']] = []
+
+        queued_dict[line['order']].append(line)
+
+    return queued_dict
+
+def timeline_to_atq(timeline, scd_dir, now):
+    """ Converts the created timeline to actual atq commands.
+
+    Args:
+        timeline (dict): A dictionary holding all timeline events.
+        scd_dir (str): The directory with SCD files.
+        now (datetime): The datetime holding the time of scheduling.
+
+    Log and backup the existing atq, remove old events and then schedule everything recent. The
+    first entry should be the currently running event, so it gets scheduled immediately.
+    """
+    backup_cmd = 'for j in $(atq | sort -k6,6 -k3,3M -k4,4 -k5,5 |cut -f 1);'\
+    'do atq |grep -P "^$j\t"; at -c "$j" | tail -n 2; done'
+
+    output = sp.check_output(backup_cmd, shell=True)
+
+    backup_time_str = now.strftime("%Y.%m.%d.%H.%M")
+    backup_dir = "{}/atq_backups".format(scd_dir)
+
+    if not os.path.exists(backup_dir):
+        os.makedirs(backup_dir)
+
+    backup_file = "{}/{}.atq".format(backup_dir, backup_time_str)
+
+    with open(backup_file, 'wb') as f:
+        f.write(output)
+
+    clear_command = "for i in `atq | awk '{print $1}'`;do atrm $i;done"
+
+    sp.call(clear_command, shell=True)
+
+    atq = []
+    first_event = True
+    for _,events in timeline.items():
+        for event in events:
+            if first_event:
+                atq.append(format_to_atq(event['time'], event['experiment'], True))
+                first_event = False
+            else:
+                atq.append(format_to_atq(event['time'], event['experiment']))
+
+    for cmd in atq:
+        sp.call(cmd, shell=True)
+
+    return sp.check_output(backup_cmd, shell=True)
```

Comment:
I was confused by this command (not used for backup here?). If it's just a sort or list of the atq commands, could maybe rename it to that

Reply:
I renamed this to something better.